packages/pubsub/src/connection-pool.js

```diff
@@ -0,0 +1,245 @@
+  var connections = Array.from(this.connections.values());
+
+  this.isOpen = false;
+  callback = callback || common.util.noop;
+
+  each(connections, function(connection, onEndCallback) {
+    connection.end(onEndCallback);
+  }, callback);
+};
+
+/**
+ * Creates a connection. This is async but instead of providing a callback
+ * a `connected` event will fire once the connection is ready.
+ */
+ConnectionPool.prototype.createConnection = function() {
+  var self = this;
+
+  this.subscription.request({
+    client: 'subscriberClient',
+    method: 'streamingPull',
+    returnFn: true
+  }, function(err, requestFn) {
+    if (err) {
+      self.emit('error', err);
+      return;
+    }
+
+    var id = uuid.v4();
+    var connection = requestFn();
+
+    connection.on('error', function(err) {
+      self.emit('error', err);
+    });
+
+    connection.on('data', function(data) {
+      arrify(data.receivedMessages).forEach(function(message) {
+        self.emit('message', self.createMessage(id, message));
+      });
+    });
+
+    connection.once('metadata', function() {
+      self.emit('connected', connection);
+    });
+
+    connection.once('close', function() {
+      self.connections.delete(id);
+
+      if (self.isOpen) {
+        self.createConnection();
+      }
+    });
+
+    if (self.isPaused) {
+      connection.pause();
+    }
+
+    connection.write({
+      subscription: self.subscription.name,
+      streamAckDeadlineSeconds: self.settings.ackDeadline / 1000
+    });
+
+    self.connections.set(id, connection);
+  });
+};
+
+/**
+ * Creates a message object for the user.
+ *
+ * @param {string} connectionId - The connection id that the message was
+ *     received on.
+ * @param {object} resp - The message response data from StreamingPull.
+ * @return {object} message - The message object.
+ */
+ConnectionPool.prototype.createMessage = function(connectionId, resp) {
+  var self = this;
+
+  var pt = resp.message.publishTime;
+  var milliseconds = parseInt(pt.nanos, 10) / 1e6;
+
+  return {
+    connectionId: connectionId,
+    ackId: resp.ackId,
+    id: resp.message.messageId,
+    data: resp.message.data,
+    attributes: resp.message.attributes,
+    publishTime: new Date(parseInt(pt.seconds, 10) * 1000 + milliseconds),
+    received: Date.now(),
+    ack: function() {
+      self.subscription.ack_(this);
+    },
+    nack: function() {
+      self.subscription.nack_(this);
+    }
+  };
+};
+
+/**
+ * Creates specified number of connections and puts pool in open state.
+ */
+ConnectionPool.prototype.open = function() {
+  for (var i = 0; i < this.settings.maxConnections; i++) {
+    this.createConnection();
+  }
+
+  this.isOpen = true;
+};
+
+/**
+ * Pauses each of the connections, causing `message` events to stop firing.
+ */
+ConnectionPool.prototype.pause = function() {
+  this.isPaused = true;
+
+  this.connections.forEach(function(connection) {
+    connection.pause();
```

Comment:
Can that happen? pause/resume aren't async, so I guess I just assumed that the mentioned scenario wasn't possible.